### ISSUE TITLE
Golang API for Multi-Cluster Env Info

### ISSUE THREAD

**Issue by simanadler** [IBM/the-mesh-for-data#144]
|   |   |
| - | - |
| Design doc | [link] |
| Parent issue | #134 |
| Required issues | # |

# Description

The M4DAppliction controller currently creates a Blueprint that spans a single cluster. It now needs to deploy to multiple clusters.  This issue is to address the additions required to inform M4DApplication what clusters are available and in what geography each cluster runs.  In the future other metadata about the cluster may also be required.

# Requirements

1. Golang API that returns a list containing a unique cluster identifier and/or cluster name, and metadata associated with the cluster.  Metadata must include geography.
2. This API should work in single and multi-cluster M4D deployments

**Comment by roee88** [comment]
```go
type ClusterMetadata struct {
  Region string
  Zone string
}

type Cluster struct {
  Name string
  Metadata ClusterMetadata
}

type ClusterManager interface {
  List() ([]Cluster, error)
}
```
Would this suffice as a starting point? @simanadler 

Only callable from coordinator cluster, no differentiation between coordinator/remote because it just lists clusters you can deploy to. Only returns clusters with metadata (cluster registration is considered as completed only when metadata is available to coordinator). Error is when we can't reach RazeeDash API. Currently using region and zone until 145 is resolved. 

@tomersolomon1 is this easily implementable?

**Comment by tomersolomon1** [comment]
Under the assumption Razee can copy a configmap from remote cluster to local cluster it should be easy to implement. Personally I haven't tried it yet, but I understand @roee88 @froesef  it's possible? Under this my idea is the follows:
1. Create a configmap holding this data in each remote cluster (during cluster registration)
2. Propagate the configmap to the coordinator cluster (using the watchkeeper)
3.  The manager goes over all these configmaps

If (2) is not possible, we can pull that configmap from the razeedash-api using graphql.
Alternatively, after connecting the remote cluster, just create that configmap in the coordinator cluster.

I'm wondering about the mapping between the `ClusterGroup` name of the cluster and the cluster-name. On one hand, we currently put only one cluster per cluster group. So it makes sense they'll have the same name or be something like `<cluster-name>-group` (like just adding a suffix).
On the other hand, maybe in the future we'd like to have multiple cluster in the same cluster group? In that case, maybe adding a field with the cluster group name to that configmap is the right way to go?
But thinking about it, grouping clusters together is a done in the coordinator cluster. So it should be different from data coming directly from remote clusters. So in the future we can have separate configmap for mapping cluster-names to their corresponding cluster group, if the mapping is not just adding a pre-defined suffix.

**Comment by roee88** [comment]
> we can pull that configmap from the razeedash-api using graphql

This is what we talked about. It needs an annotation on the configmap with level set to capture everything instead of just status.

**Comment by tomersolomon1** [comment]
@roee88 Where would that configmap would be created?

**Comment by shlomitk1** [comment]
@roee88 @simanadler I believe we need this information for a single cluster config as well. Cluster name is less important, can be "local" or something alike, but a mapping to geography is essential in order to apply governance policies correctly.

**Comment by roee88** [comment]
Known :)

**Comment by roee88** [comment]
Already implemented with #176 its follow-ups
